### ISSUE TITLE
setTimeout doesn't work with the sleep command

### ISSUE THREAD

**Issue by neskhodovskiy** [phpseclib/phpseclib#1440]
Hello, thank you for the wonderful library and all the work you do with phpseclib. 

I am curious about the following case.

When I'm setting timeout for some command like 'ping' it works as expected.

When I'm setting timeout and call 'sleep', command waits the exact number of seconds passed to sleep, instead of respecting the timeout set to SSH2.

The code:

```php
<?php
use phpseclib\Crypt\RSA;
use phpseclib\Net\SSH2;

$key = new RSA();
$key->loadKey(file_get_contents('private-key.txt'));

$ssh = new SSH2('127.0.0.1');
if (!$ssh->login('username', $key)) {
    exit('Login Failed');
}

$ssh->setTimeout(3);

$start = microtime(true);
$ssh->exec('ping google.com');
$elapsed = microtime(true) - $start;

if ($ssh->isTimeout()) {
   echo sprintf('Command timed out in %01.2f seconds', $elapsed);
} else {
   echo sprintf('Command completed in %01.2f seconds', $elapsed);
}
```

The code above returns as expected:

> Command timed out in 3.12 seconds

However:

```php
<?php
use phpseclib\Crypt\RSA;
use phpseclib\Net\SSH2;

$key = new RSA();
$key->loadKey(file_get_contents('private-key.txt'));

$ssh = new SSH2('127.0.0.1');
if (!$ssh->login('username', $key)) {
    exit('Login Failed');
}

$ssh->setTimeout(3);

$start = microtime(true);
$ssh->exec('sleep 10');
$elapsed = microtime(true) - $start;

if ($ssh->isTimeout()) {
   echo sprintf('Command timed out in %01.2f seconds', $elapsed);
} else {
   echo sprintf('Command completed in %01.2f seconds', $elapsed);
}
```
I am expecting the command to time out after 3 seconds. But the second code runs for 10 seconds instead of 3 and outputs

> Command timed out in 10.11 seconds

What am I missing here?

**Comment by terrafrost** [comment]
I'm able to reproduce the issue.

From the logs:

```
-> NET_SSH2_MSG_CHANNEL_REQUEST (since last: 0.0004, network: 0.0001s)
00000000  00:00:00:00:00:00:00:04:65:78:65:63:01:00:00:00  ........exec....
00000010  08:73:6c:65:65:70:20:31:30                       .sleep 10

<- NET_SSH2_MSG_CHANNEL_WINDOW_ADJUST (since last: 0.1306, network: 0.0001s)
00000000  00:00:00:01:00:20:00:00                          ..... ..

<- NET_SSH2_MSG_CHANNEL_SUCCESS (since last: 0.0006, network: 0.0001s)
00000000  00:00:00:01                                      ....

-> NET_SSH2_MSG_CHANNEL_EOF (since last: 2.7624, network: 0.0003s)
00000000  00:00:00:00                                      ....

-> NET_SSH2_MSG_CHANNEL_CLOSE (since last: 0.0007, network: 0.0001s)
00000000  00:00:00:00                                      ....

<- NET_SSH2_MSG_CHANNEL_REQUEST (since last: 7.2396, network: 0.0002s)
00000000  00:00:00:01:00:00:00:0b:65:78:69:74:2d:73:74:61  ........exit-sta
00000010  74:75:73:00:00:00:00:00                          tus.....

<- NET_SSH2_MSG_CHANNEL_CLOSE (since last: 0.0007, network: 0.0001s)
00000000  00:00:00:01                                      ....
```
Per that phpseclib does try to close the channel after ~3s. The problem is that the server isn't acknowledging the close request until ~7s after it was sent (ie. after which ~10s has cumulatively passed).

Unfortunately, I'm not really sure what can be done about this. phpseclib could be updated so that it doesn't wait for the close request to be acknowledged by the server but then what? If you just want to issue that one command then it's not a problem but if you do want to issue other commands then it becomes more complicated.

In phpseclib 4.0 I want to rewrite SSH2 so that it supports multiple channels. At that point phpseclib wouldn't need to wait for the close confirmation - every `exec` would be on it's own unique channel. But that's not an option for phpseclib 1.0 - 3.0. For these we'd just need to wait for that close confirmation before the the next command could be ran.

Meh. I'm sure that's not the answer you were hoping to hear :(

**Comment by neskhodovskiy** [comment]
No worries, thank you for at least looking into the issue. I guess I'll wait for the future versions then - good luck with the releases!

**Comment by terrafrost** [comment]
I was playing around with this and found a way to work around this sorta:

```php
$ssh->exec('nohup sleep 10 > /dev/null 2>&1 &')
```

**Comment by neskodi** [comment]
> I was playing around with this and found a way to work around this sorta:
> 
> ```
> $ssh->exec('nohup sleep 10 > /dev/null 2>&1 &')
> ```

Putting the command to run in background might be a solution indeed for a number of cases. In this case, the timeout setting doesn't make much sense at all - the `exec` will return immediately anyway. However, thanks for keeping thinking about it.

**Comment by neskodi** [comment]
Just in case, the following way of setting a timeout will also work perfectly with `sleep`:

`$ssh->exec('timeout 5 sleep 10');`

As you can see, it relies on having the `timeout` utility from GNU coreutils available on the remote system, so YMMV. AFAIK it is available on most modern linux distributions.

**Comment by splitice** [comment]
I believe this related to an issue I'm seeing as well. We had a faulty linecard at a PoP (fixed now) and it caused all connections to hang. The PHP workers did got stuck in phpseclib.

It looks like they were waiting on a response to the channel closing. I can't force it to happen again, however it's the kind of thing that can easily happen and having it lock up a PHP worker is extremely bad.

It looks like before closing the channel `$this->curTimeout = 0;` therefore the operation never times-out. In the case the other end is hung / dead this is a very bad idea.

**Comment by terrafrost** [comment]
@splitice - are you explicitly setting the timeout to 0 with `$ssh->setTimeout(0)` or is that what it's ending up as?

I will say that in theory it is possible for you to have a timeout of, say, 10s (eg. `$ssh->setTimeout(10)`) and for `$ssh->curTimeout` to just wind up at 0, exactly, but it's very unlikely.

From `_get_channel_packet`:

```php
                    if ($this->curTimeout < 0) {
                        $this->is_timeout = true;
                        return true;
                    }
...
                    $elapsed = microtime(true) - $start;
                    $this->curTimeout-= $elapsed;
```
I can look into making this theoretical issue can't _ever_ happen but you'd have to have extremely bad luck for that to just randomly happen to you.

`microtime(true)` "returns the current Unix timestamp with microseconds". One microsecond is 0.000001 seconds. So you'd have to match the remaining time of `$this->curTimeout` down to the microsecond for this theoretical issue to occur. idk it's def possible it just doesn't seem probably idk.

Or am I misunderstanding your comment?

**Comment by splitice** [comment]
@terrafrost I'm running with `$ssh->setTimeout(10)`.

You can find the timeout being set in locations like:
 - https://github.com/phpseclib/phpseclib/blob/2.0/phpseclib/Net/SSH2.php#L3897
 - https://github.com/phpseclib/phpseclib/blob/2.0/phpseclib/Net/SSH2.php#L4126

Which in turn results in a `socket_select` where tv_sec is NULL.  And to quote the PHP manual:

> If tv_sec is NULL stream_select() can block indefinitely, returning only when an event on one of the watched streams occurs (or if a signal interrupts the system call).

**Comment by terrafrost** [comment]
@splitice - thank you for the clarification.

`$this->curTimeout = 0;` is being set in those places because if the channel is not closed due to timing out prematurely then that will interfere with the channel's being reused. 

In theory, in the future, phpseclib will support multiple channels so this won't be as necessary in the future.

> Not entirely related but there really should be a flag on the SSH2 stream that marks the instance as dead to prevent further attempts to use the stream if the channel fails to close. More of a nice to have though.

A lot of the public facing methods (eg. `$ssh->exec()`) do this at the beginning:

```php
        if (!$this->isAuthenticated()) {
            return false;
        }
```
Here's the definition of `isAuthenticated`:

```php
    function isAuthenticated()
    {
        return (bool) ($this->bitmap & self::MASK_LOGIN);
    }
```
And whenever there's an error `$this->bitmap` is set to 0.

Seems like that's pretty much doing what you're describing?

**Comment by splitice** [comment]
> $this->curTimeout = 0; is being set in those places because if the channel is not closed due to timing out prematurely then that will interfere with the channel's being reused.

There won't be a channel to be re-used if the stream is blocked indefinitely however. The PHP process will never return to processing from socket_select.

**Comment by terrafrost** [comment]
@splitice - in the scenario you're describing you're correct. However, that is not the only scenario that phpseclib has to consider.

Let me mull on this for a few days.

**Comment by splitice** [comment]
@terrafrost any progress?

**Comment by terrafrost** [comment]
Here's one idea I had:

https://github.com/terrafrost/phpseclib/commit/9651d075e3e023001e4581ef4b86fc04e1025593

But tbh I'm not sure it helps your use case any more than the current behavior does and even if it did I could see it causing hard to foresee problems in other virtually one-off scenarios.

In doing a cost benefit analysis...  the cost is pretty low (but not non-zero) but I see the benefit as being pretty low as well (and that's even assuming that there's a benefit at all; like I said, I'm not sure the change actually helps you).

The net effect is that my enthusiasm is pretty muted.

I guess lmk what you think of the change and I'll see how I feel.

**Comment by splitice** [comment]
I've been running a non infinite timeout since mid may. No issues for my use case since we dispose of the SSH instance on critical error.

**Comment by nunoperalta** [comment]
This is an interesting issue. I was going to report something like this.

I'm running a "docker-compose run" process, with a `setTimeout(8 hours)`, but if the process takes 9 hours to run, it will wait the 9 hours.
After reading the above thread, I understand why.

Might try using the `timeout X .....` way, so!
